src/vision/status_quo/barbara_carefully_dismisses_embedded_future.md

```diff
@@ -0,0 +1,210 @@
+so that the library can be compatible with the async Rust ecosystem. To do
+this, she decides to make the buffer printing API return a Future. She starts
+with a skeleton:
+
+```rust
+/// Passes `buffer` to the kernel, and prints it to the console. Returns a
+/// future that returns `buffer` when the print is complete. The caller must
+/// call kernel_ready_for_callbacks() when it is ready for the future to return. 
+fn print_buffer(buffer: &'static mut [u8]) -> PrintFuture {
+    // TODO: Set the callback
+    // TODO: Tell the kernel to print `buffer`
+}
+
+struct PrintFuture;
+
+impl core::future::Future for PrintFuture {
+    type Output = &'static mut [u8];
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        // TODO: Detect when the print is done, retrieve `buffer`, and return
+        // it.
+    }
+}
+```
+
+Note: All error handling is omitted to keep things understandable.
+
+Barbara begins to implement `print_buffer`:
+
+```rust
+fn print_buffer(buffer: &'static mut [u8]) -> PrintFuture {
+    kernel_set_print_callback(callback);
+    kernel_start_print(buffer);
+    PrintFuture {}
+}
+
+// New! The callback the kernel calls.
+extern fn callback() {
+    // TODO: Wake up the currently-waiting PrintFuture.
+}
+```
+
+So far so good. Barbara then works on `poll`:
+
+```rust
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        if kernel_is_print_done() {
+            Poll::Ready(kernel_get_buffer_back())
+        }
+        Poll::Pending
+    }
+```
+
+Of course, there's something missing here. How does the callback wake the
+`PrintFuture`? She needs to store the
+[`Waker`](https://doc.rust-lang.org/stable/core/task/struct.Waker.html)
+somewhere! Barbara puts the `Waker` in a global variable so the callback can
+find it (this is fine because the app is single threaded and callbacks do NOT
+interrupt execution the way Unix signals do):
+
+```rust
+static mut PRINT_WAKER: Option<Waker> = None;
+
+extern fn callback() {
+    if let Some(waker) = unsafe { PRINT_WAKER.as_ref() } {
+        waker.wake_by_ref();
+    }
+}
+```
+
+She then modifies `poll` to set `PRINT_WAKER`:
+
+```rust
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        if kernel_is_print_done() {
+            return Poll::Ready(kernel_get_buffer_back());
+        }
+        unsafe { PRINT_WAKER = Some(cx.waker()); }
+        Poll::Pending
+    }
+```
+
+At this point, Barbara starts to think about the size impact of these changes.
+`PRINT_WAKER` is stored in `.bss`, which occupies space in RAM but not flash. It
+is two words in size. It points to a
+[`RawWakerVTable`](https://doc.rust-lang.org/stable/core/task/struct.RawWakerVTable.html)
+that is provided by the executor. Considering that this code is designed to run
+without `alloc`, `RawWakerVTable` seems awfully wasteful. `drop` is a no-op,
```

Comment:
It would be great to talk about the impact of this waste. `RawWakerVTable` is only 4 words big. Probably many reading this story won't understand why 4 words is considered wasteful.

Reply:
(chiming in from Tock here, so ignore if not useful) @jrvanwhy: did you mean that the stable itself (which is 4 words in flash) is wasteful or the 2 words in memory---seems wasteful because without alloc we actually could know statically exactly what concrete type it would be, and also because we need these 2 words for _each_ instance of async, rather than once for any program that uses futures?

Maybe both?

My interpretation had been the 2 words in memory, but @rylev's comment highlighted to me that I actually am not sure which this is referring to.

Reply:
If there's a single copy of `RawWakerVTable` in flash, that's not bad. If there's 1 copy in RAM for each executor instance, that's pretty significant.

When I wrote the executor, I tried to avoid dynamic dispatch so the compiler could perform inlining. This meant the executor depended on the future's type, like:

```rust
pub struct Task<F: Future> {
    future: Option<F>,
}
```

Because of this, I needed the RawWakerVTable to be specific to the type of Future. I implemented this by putting it directly into the executor, i.e.:

```rust
pub struct Task<F: Future> {
    future: Option<F>,
    vtable: RawWakerVTable,
}
```

As a result, I had a `RawWakerVTable` instance in RAM for each `Task`.

Because `future` needs to be mutable, the `Task` gets placed in `.data` in RAM. As a result, the vtable ends up in RAM, and there's one per `Task`.

In retrospect, I could have broken the executor into two parts:

```rust
// Instantiated by client code, lives in flash.
pub struct TaskVTable<F: Future> {
    vtable: RawWakerVTable
}

// Instantiated by client code, lives in RAM
pub struct Task<F: future> {
    future: Option<F>,
    vtable: &'static TaskVTable<F>,
}
```

This would reduce the RAM overhead to 1 word per `Task` and move the overall vtable into RAM. I expect you could do the same with future combinators (although most combinators probably don't need their own `RawWakerVTable` anyway).

If you use the `Locator` trait from my async API design, we could remove the `vtable` member from `Task` and replace it with a zero-sized type.